### DISCUSSION TITLE
Need help about my neper

### DISCUSSION THREAD

**Discussion by changhaoliu-UOM** [neperfepx/neper#929] [Q&A] [answered]
Hi All

I want to generate a model with a given cell size distribution (volume).
The volume distribution is here. And the code I used is also attached below.
[volume.txt](https://github.com/user-attachments/files/17460354/volume.txt)
neper -T -n from_morpho                                                       \
           -morpho "size:file(volume.txt),1-sphericity:lognormal(0.145,0.03)"                                     \
            -dim 3                                                               \
           -domain "cube(2154.3,2154.3,2154.3)"                                            \
           -o mytess

But I cannot get a model owing a close distribution and the total number of grains is not correct.
I am not sure if there is a mistake in my txt or in the code.

best

Changhao

**Comment by rquey** [comment, marked as answer]
The correct command is
```
neper -T -n from_morpho -morpho "size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03)" -dim 3 -domain "cube(2154.3,2154.3,2154.3)" -o mytess
```

`-morpho "size:file(volume.txt)"` loads discrete values for the cells (which does not make sense with your file), while `-morpho "size:custom(volume.txt)"` reads a custom frequency distribution from the file.


I've also made a fix in version 4.10.2-1.

**Reply by changhaoliu-UOM** [reply]
Thank you very much.

I tried the code as your advice. However it always stops in the middle. I am not sure if there is a problem with my file.

best
Changhao

**Reply by rquey** [reply]
Have you actually updated?

**Reply by changhaoliu-UOM** [reply]
Yes, but after a very short time, it aborted. Is there any requirement to the file? For example, the sum of the volumes in the file is not the total domain's volume. Does it matter? Or is there anything wrong? I tried both counts and frequency but neither work.
CODE:neper -T -n from_morpho -morpho "size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03)" -dim 3 -domain "cube(2154.3,2154.3,2154.3)" -o mytess
File:
[volume.txt](https://github.com/user-attachments/files/17519522/volume.txt)

**Reply by rquey** [reply]
On my laptop, it took a while but it did it (again, with `4.10.2-1`, which has the fix):
```
$ neper -T -n from_morpho -morpho "size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03)" -dim 3 -domain "cube(2154.3,2154.3,2154.3)" -o mytess                                                                    

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.10.2-1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 16 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2024, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n from_morpho -morpho
         size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03) -dim
         3 -domain cube(2154.3,2154.3,2154.3) -o mytess
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellation...
Info   :     [i] Parsing file `volume.txt'...
Info   :     [i] Parsed file `volume.txt'.
Info   :       . Number of cells: 2943
Info   :   - Setting seeds... 100%
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =1.095870019
Info   :     > Iteration   111551629: fmin=1.022034153 f=1.022155726
r     3424386: fmin==0.001512636 f==0.001512638
Info   :     > Final solution  : f   =0.001512636 (3424386 iterations)
Info   :     > Reached `eps' criterion.
Info   : Generating crystal orientations...
Info   : Writing results...
Info   :     [o] Writing file `mytess.tess'...
Info   :     [o] Wrote file `mytess.tess'.
Info   : Elapsed time: 21045.395 secs.
========================================================================
```
I suggest you use `-morphooptistop eps<1e-4` or similar (instead of `eps<1e-6`) to go faster.

![img](https://github.com/user-attachments/assets/a5d44dc3-eed0-42b7-9c4c-ade789bb5d5c)

**Comment by changhaoliu-UOM** [comment]
Thank you so much. However, I copied your code and used the same txt file. But i only got 1025 cells which is not so correct. Your result owning 2943 cells is almost correct. Did you make modifications on the txt file? Or is there any possible mistakes I made? 
```bash
 $ neper -T -n from_morpho -morpho "size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03)" -dim 3 -domain "cube(2154.3,2154.3,2154.3)" -o mytess

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.10.2-1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 16 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2024, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/changhao/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n from_morpho -morpho
         size:custom(volume.txt),1-sphericity:lognormal(0.145,0.03) -dim
         3 -domain cube(2154.3,2154.3,2154.3) -o mytess
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellation...
Info   :     [i] Parsing file `volume.txt'...
Info   :     [i] Parsed file `volume.txt'.
Info   :       . Number of cells: 1025
Info   :   - Setting seeds... 100%
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =1.087495701
Info   :     > Iteration   119093: fmin=0.894542185 f=0.894542185
Info   :     > Final solution  : f   =0.894542185 (1193 iterations)
Warning:     > Stopping on Ctrl+C invocation.
Info   : Generating crystal orientations...
Info   : Writing results...
```
